### ISSUE TITLE
Matrix determinant calculation is too aggressive with simplification

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#11082]
You can see [here](https://github.com/sympy/sympy/blob/ef47677483b2f29d0b8e6a0eb45de72b2e34477d/sympy/matrices/matrices.py#L2612) that the bareis determinant method calls `cancel` in order to cancel out fractions. But if the matrix in question has large expressions, this does too much, and can slow things down. 

See for instance https://stackoverflow.com/questions/37026935/speeding-up-computation-of-symbolic-determinant-in-sympy, where it was faster to compute the general formula for the determinant and substitute values in than to compute `M.det()` directly.

**Comment by krishnacharya** [comment]
@asmeurer I was thinking of implementing the change by always calculating the general formula for the n*n Matrix then substituting the specific values of A00 A01 .. , should i proceed in this way

**Comment by baoqchau** [comment]
Hello @asmeurer. I want to fix this. Do you have any suggestion how to do it?

**Comment by asmeurer** [comment]
It would need to do that conditionally. For a large sparse matrix, for instance, that would be much slower.

**Comment by siefkenj** [comment]
For matrices with numbers, it seems that up to 4x4, it's faster to use the general formula, and after that, Bareiss is faster.

**Comment by asmeurer** [comment]
Same issue occurs for computing inverses of general symbolic matrices.

**Comment by oscarbenjamin** [comment]
Are you sure it's the same issue or is it perhaps a newer issue (#19920)?

**Comment by asmeurer** [comment]
Ah, I was trying to find an issue for inverses specifically. I suspect they both have similar causes, since in each case computing the general formula without trying to simplify it is better for a fully symbolic matrix.

**Comment by oscarbenjamin** [comment]
Yes, but #19920 is a newer issue caused by the changes in 1.6. Many of those changes were reverted but that one remains (I'll push the milestone to 1.7 but it's a release blocker)

**Comment by oscarbenjamin** [comment]
Also note that there is a new (not public yet) matrix implementation in `sympy/polys/domainmatrix` which is much faster for many things e.g.:
```julia
In [130]: M = Matrix([symbols('x:9')]).reshape(3, 3)                                                                                           

In [131]: %time ok=M.inv()                                                                                                                     
CPU times: user 1.21 s, sys: 7.04 ms, total: 1.22 s
Wall time: 1.23 s

In [132]: from sympy.polys.domainmatrix import DomainMatrix                                                                                    

In [133]: M2 = DomainMatrix.from_list_sympy(3, 3, M.tolist()).to_field()                                                                       

In [134]: %time ok=M2.inv()                                                                                                                    
CPU times: user 78.1 ms, sys: 569 µs, total: 78.7 ms
Wall time: 79.3 ms
```

**Comment by asmeurer** [comment]
I don't know how practical this is, but there has been some recent work on representing the general determinant formula with fewer than n! terms https://cp4space.hatsya.com/2022/07/02/a-combinatorial-proof-of-houstons-identity/
